src/OpenTelemetry.Extensions.Owin/DiagnosticsMiddleware.cs

```diff
@@ -0,0 +1,158 @@
+        private readonly Context context = new Context();
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="DiagnosticsMiddleware"/> class.
+        /// </summary>
+        /// <param name="next">An optional pointer to the next component</param>
+        public DiagnosticsMiddleware(OwinMiddleware next)
+            : base(next)
+        {
+        }
+
+        /// <inheritdoc />
+        public override async Task Invoke(IOwinContext owinContext)
+        {
+            try
+            {
+                this.BeginRequest(owinContext);
+                await this.Next.Invoke(owinContext).ConfigureAwait(false);
+                this.RequestEnd(owinContext, null);
+            }
+            catch (Exception ex)
+            {
+                this.RequestEnd(owinContext, ex);
+            }
+        }
+
+        // Based on https://github.com/dotnet/aspnetcore/blob/v5.0.0-rc.2.20475.17/src/Hosting/Hosting/src/Internal/HostingApplicationDiagnostics.cs#L37
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private void BeginRequest(IOwinContext owinContext)
+        {
+            if (OwinExtensionsEventSource.Log.IsEnabled())
+            {
+                this.context.EventLogEnabled = true;
+            }
+
+            if (this.diagnosticListener.IsEnabled() && this.diagnosticListener.IsEnabled(ActivityName, owinContext))
+            {
+                this.context.Activity = this.StartActivity(owinContext, out var hasDiagnosticListener);
+                this.context.HasDiagnosticListener = hasDiagnosticListener;
+            }
+        }
+
+        // Based on https://github.com/dotnet/aspnetcore/blob/v5.0.0-rc.2.20475.17/src/Hosting/Hosting/src/Internal/HostingApplicationDiagnostics.cs#L89
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private void RequestEnd(IOwinContext owinContext, Exception exception)
+        {
+            var activity = this.context.Activity;
+            // Always stop activity if it was started
+            if (activity != null)
+            {
+                this.StopActivity(owinContext, activity, this.context.HasDiagnosticListener);
+            }
+
+            if (this.context.EventLogEnabled && exception != null)
+            {
+                // Non-inline
+                OwinExtensionsEventSource.Log.UnhandledException();
+            }
+        }
+
+        [MethodImpl(MethodImplOptions.NoInlining)]
+        private Activity StartActivity(IOwinContext owinContext, out bool hasDiagnosticListener)
+        {
+            hasDiagnosticListener = false;
+
+            var activity = new Activity(ActivityName);
+
+            // Based on https://github.com/microsoft/ApplicationInsights-dotnet/blob/2.15.0/WEB/Src/Web/Web/Implementation/RequestTrackingExtensions.cs#L41
+            if (!activity.Extract(owinContext.Request.Headers))
```

Comment:
We need to move on to `Propagagators.DefaultTextMapPropagator` approach.

Reply:
Can you please point me to some examples?

Reply:
https://github.com/open-telemetry/opentelemetry-dotnet/blob/master/examples/MicroserviceExample/Utils/Messaging/MessageSender.cs